File: project/.sdk/tm/py/feature/test_feature.py
# ProjectName SDK test feature

from __future__ import annotations
import random

from utility.voxgig_struct import voxgig_struct as vs
from feature.base_feature import ProjectNameBaseFeature


class ProjectNameTestFeature(ProjectNameBaseFeature):
    def __init__(self):
        super().__init__()
        self.version = "0.0.1"
        self.name = "test"
        self.active = True
        self.client = None
        self.options = None

    def init(self, ctx, options):
        self.client = ctx.client
        self.options = options

        entity = vs.getprop(options, "entity")
        if not isinstance(entity, dict):
            entity = {}

        self.client.mode = "test"

        # Ensure entity ids are correct.
        vs.walk(entity, lambda key, val, parent, path: (
            val.__setitem__("id", key) if len(path) == 2 and isinstance(val, dict) and key is not None else None,
            val
        )[-1])

        test_self = self

        def test_fetcher(fctx, _fullurl, _fetchdef):
            def respond(status, data, extra=None):
                out = {
                    "status": status,
                    "statusText": "OK",
                    "json": lambda: data,
                    "body": "not-used",
                }
                if isinstance(extra, dict):
                    for k, v in extra.items():
                        out[k] = v
                return out, None

            op = fctx.op
            entmap = vs.getprop(entity, op.entity)
            if not isinstance(entmap, dict):
                entmap = {}

            # For single-entity ops (load, remove) with an empty explicit
            # match, fall back to the id the entity client already knows from a
            # prior create/load (in fctx.match / fctx.data). Mirrors the TS
            # mock where param() resolves the id from that accumulated state.
            def _resolve_match(explicit):
                if isinstance(explicit, dict) and len(explicit) > 0:
                    return explicit
                for src in (getattr(fctx, "match", None), getattr(fctx, "data", None)):
                    v = vs.getprop(src, "id") if src is not None else None
                    if v is not None and v != "__UNDEFINED__":
                        return {"id": v}
                return {}

            if op.name == "load":
                args = test_self.build_args(fctx, op, _resolve_match(fctx.reqmatch))
                found = vs.select(entmap, args)
                ent = vs.getelem(found, 0)
                if ent is None:
                    return respond(404, None, {"statusText": "Not found"})
                vs.delprop(ent, "$KEY")
                out = vs.clone(ent)
                return respond(200, out)

            elif op.name == "list":
                args = test_self.build_args(fctx, op, fctx.reqmatch)
                found = vs.select(entmap, args)
                if found is None:
                    return respond(404, None, {"statusText": "Not found"})
                if isinstance(found, list):
                    for item in found:
                        vs.delprop(item, "$KEY")
                out = vs.clone(found)
                return respond(200, out)

            elif op.name == "update":
                # Match the existing entity by id only (or its alias). reqdata
                # also contains the new field values, which would otherwise
                # cause select to filter out the entity we want to update.
                # When reqdata has no id, fall back to the id the entity
                # client carries from a prior create/load (in fctx.match /
                # fctx.data), mirroring the TS mock where param(ctx,'id')
                # resolves from accumulated state.
                update_match = {}
                if isinstance(fctx.reqdata, dict):
                    if "id" in fctx.reqdata:
                        update_match["id"] = fctx.reqdata["id"]
                    alias_map = getattr(op, "alias_map", None)
                    if alias_map is not None:
                        alias_id = vs.getprop(alias_map, "id")
                        if alias_id is not None and alias_id in fctx.reqdata:
                            update_match[alias_id] = fctx.reqdata[alias_id]
                if not update_match:
                    update_match = _resolve_match({})
                args = test_self.build_args(fctx, op, update_match)
                found = vs.select(entmap, args)
                ent = vs.getelem(found, 0)
                if ent is None and isinstance(entmap, dict):
                    for e in entmap.values():
                        if isinstance(e, dict):
                            ent = e
                            break
                if ent is None:
                    return respond(404, None, {"statusText": "Not found"})
                if isinstance(ent, dict):
                    reqdata = fctx.reqdata
                    if reqdata is not None:
                        for k, v in reqdata.items():
                            ent[k] = v
                vs.delprop(ent, "$KEY")
                out = vs.clone(ent)
                return respond(200, out)

            elif op.name == "remove":
                args = test_self.build_args(fctx, op, _resolve_match(fctx.reqmatch))
                found = vs.select(entmap, args)
                ent = vs.getelem(found, 0)
                # Remove only the first matched entity. If nothing matches,
                # succeed as a no-op rather than erroring.
                if isinstance(ent, dict):
                    eid = vs.getprop(ent, "id")
                    vs.delprop(entmap, eid)
                return respond(200, None)

            elif op.name == "create":
                test_self.build_args(fctx, op, fctx.reqdata)
                eid = fctx.utility.param(fctx, "id")
                if eid is None:
                    eid = "%04x%04x%04x%04x" % (
                        random.randint(0, 0xFFFF), random.randint(0, 0xFFFF),
                        random.randint(0, 0xFFFF), random.randint(0, 0xFFFF))

                ent = vs.clone(fctx.reqdata)
                if isinstance(ent, dict):
                    ent["id"] = eid
                    if isinstance(eid, str):
                        entmap[eid] = ent
                    vs.delprop(ent, "$KEY")
                    out = vs.clone(ent)
                    return respond(200, out)
                return respond(200, ent)

            return respond(404, None, {"statusText": "Unknown operation"})

        ctx.utility.fetcher = test_fetcher

    def build_args(self, ctx, op, args):
        opname = op.name

        # Get last point from config.
        points = vs.getpath(ctx.config, "entity." + ctx.entity.get_name() + ".op." + opname + ".points")
        point = vs.getelem(points, -1)

        # Get required params.
        params_path = vs.getpath(point, "args.params")
        reqd_params = vs.select(params_path, {"reqd": True})
        reqd = vs.transform(reqd_params, ["`$EACH`", "", "`$KEY.name`"])

        qand = []
        q = {"`$AND`": qand}

        if args is not None:
            keys = vs.keysof(args)
            if keys is not None:
                for key in keys:
                    is_id = (key == "id")
                    selected = vs.select(reqd, key)
                    is_reqd = not vs.isempty(selected)

                    if is_id or is_reqd:
                        v = ctx.utility.param(ctx, key)
                        ka = None
                        if op.alias is not None:
                            ka = vs.getprop(op.alias, key)

                        qor = [{key: v}]
                        if ka is not None and isinstance(ka, str):
                            qor.append({ka: v})

                        qand.append({"`$OR`": qor})

        q["`$AND`"] = qand

        if ctx.ctrl.explain is not None:
            ctx.ctrl.explain["test"] = {"query": q}

        return q

File: project/.sdk/tm/rb/feature/test_feature.rb
# ProjectName SDK test feature

require_relative '../utility/struct/voxgig_struct'
require_relative 'base_feature'

class ProjectNameTestFeature < ProjectNameBaseFeature
  def initialize
    super
    @version = "0.0.1"
    @name = "test"
    @active = true
    @client = nil
    @options = nil
  end

  def init(ctx, options)
    @client = ctx.client
    @options = options

    entity = VoxgigStruct.getprop(options, "entity")
    entity = {} unless entity.is_a?(Hash)

    @client.mode = "test"

    # Ensure entity ids are correct.
    VoxgigStruct.walk(entity) do |key, val, parent, path|
      if path.length == 2 && val.is_a?(Hash) && key
        val["id"] = key
      end
      val
    end

    test_self = self

    test_fetcher = ->(fctx, _fullurl, _fetchdef) {
      respond = ->(status, data, extra) {
        out = {
          "status" => status,
          "statusText" => "OK",
          "json" => -> { data },
          "body" => "not-used",
        }
        extra&.each { |k, v| out[k] = v }
        return out, nil
      }

      op = fctx.op
      entmap = VoxgigStruct.getprop(entity, op.entity)
      entmap = {} unless entmap.is_a?(Hash)

      if op.name == "load"
        args = test_self.build_args(fctx, op, fctx.reqmatch)
        found = VoxgigStruct.select(entmap, args)
        ent = VoxgigStruct.getelem(found, 0)
        return respond.call(404, nil, { "statusText" => "Not found" }) unless ent
        VoxgigStruct.delprop(ent, "$KEY")
        out = VoxgigStruct.clone(ent)
        respond.call(200, out, nil)

      elsif op.name == "list"
        args = test_self.build_args(fctx, op, fctx.reqmatch)
        found = VoxgigStruct.select(entmap, args)
        return respond.call(404, nil, { "statusText" => "Not found" }) unless found
        if found.is_a?(Array)
          found.each { |item| VoxgigStruct.delprop(item, "$KEY") }
        end
        out = VoxgigStruct.clone(found)
        respond.call(200, out, nil)

      elsif op.name == "update"
        args = test_self.build_args(fctx, op, fctx.reqdata)
        found = VoxgigStruct.select(entmap, args)
        ent = VoxgigStruct.getelem(found, 0)
        return respond.call(404, nil, { "statusText" => "Not found" }) unless ent
        if ent.is_a?(Hash) && fctx.reqdata
          fctx.reqdata.each { |k, v| ent[k] = v }
        end
        VoxgigStruct.delprop(ent, "$KEY")
        out = VoxgigStruct.clone(ent)
        respond.call(200, out, nil)

      elsif op.name == "remove"
        args = test_self.build_args(fctx, op, fctx.reqmatch)
        found = VoxgigStruct.select(entmap, args)
        ent = VoxgigStruct.getelem(found, 0)
        return respond.call(404, nil, { "statusText" => "Not found" }) unless ent
        if ent.is_a?(Hash)
          id = VoxgigStruct.getprop(ent, "id")
          VoxgigStruct.delprop(entmap, id)
        end
        respond.call(200, nil, nil)

      elsif op.name == "create"
        test_self.build_args(fctx, op, fctx.reqdata)
        id = fctx.utility.param.call(fctx, "id")
        id ||= "%04x%04x%04x%04x" % [rand(0x10000), rand(0x10000), rand(0x10000), rand(0x10000)]

        ent = VoxgigStruct.clone(fctx.reqdata)
        if ent.is_a?(Hash)
          ent["id"] = id
          entmap[id.to_s] = ent if id.is_a?(String)
          VoxgigStruct.delprop(ent, "$KEY")
          out = VoxgigStruct.clone(ent)
          return respond.call(200, out, nil)
        end
        respond.call(200, ent, nil)

      else
        respond.call(404, nil, { "statusText" => "Unknown operation" })
      end
    }

    ctx.utility.fetcher = test_fetcher
  end

  def build_args(ctx, op, args)
    opname = op.name
    points = VoxgigStruct.getpath(ctx.config, "entity.#{ctx.entity.get_name}.op.#{opname}.points")
    point = VoxgigStruct.getelem(points, -1)

    params_path = VoxgigStruct.getpath(point, "args.params")
    reqd_params = VoxgigStruct.select(params_path, { "reqd" => true })
    reqd = VoxgigStruct.transform(reqd_params, ["`$EACH`", "", "`$KEY.name`"])

    qand = []
    q = { "`$AND`" => qand }

    if args
      keys = VoxgigStruct.keysof(args)
      if keys
        keys.each do |key|
          is_id = (key == "id")
          selected = VoxgigStruct.select(reqd, key)
          is_reqd = !VoxgigStruct.isempty(selected)

          if is_id || is_reqd
            v = ctx.utility.param.call(ctx, key)
            ka = op.alias_map ? VoxgigStruct.getprop(op.alias_map, key) : nil

            qor = [{ key => v }]
            qor << { ka => v } if ka.is_a?(String)

            qand << { "`$OR`" => qor }
          end
        end
      end
    end

    q["`$AND`"] = qand
    ctx.ctrl.explain["test"] = { "query" => q } if ctx.ctrl.explain

    q
  end
end
